### PR TITLE
[TextViewer] Add hamburger menu to FocusManager and Menu key

### PR DESCRIPTION
**Pull request by Commodore64user**
### what's new

* Add hamburger menu to FocusManager and <kbd>Menu</kbd> key

another stray widget that needs its menu mapped to Menu key. borrowing code from @NiLuJe to fix focusmanager shenanigans.

I might be pushing the functionality of `Device:useDPadAsActionKeys()` to mean, has a dedicated <kbd>Menu</kbd> button... await judgment there.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/12365)

### CODE REVIEW

File: frontend/ui/widget/textviewer.lua [hunk 1]
```diff
@@ -272,6 +274,16 @@ function TextViewer:init(reinit)
         show_parent = self,
     }
 
+    -- NT: add titlebar.left_button (hamburger menu) to FocusManager.
+    if Device:hasDPad() and not Device:useDPadAsActionKeys() then
```
Comment (Frenzie) [inline frontend/ui/widget/textviewer.lua:278]: What's the purpose of the complex hard to follow conditions instead of just keeping it simple?

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:278]: In layman's terms, don't add it on kindle or other device where there might be a Menu button. Although come to think of it, it could also be ` hasDPad and hasFewKeys`

Comment (Frenzie) [reply frontend/ui/widget/textviewer.lua:278]: But why not? Why make things difficult and complex?

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:278]: Because having things on older kindles that don’t really need to be there makes it slower, particularly k2 and k3. Yes it is just one extra item to cycle through, but that one extra key press takes time, furthermore reduces performance so, i’d rather not have it.

if i was selfish, i would only add the button mapping. But since i am in the business of making it better to all, i also want other to be able to have it.

Comment (Frenzie) [reply frontend/ui/widget/textviewer.lua:278]: > Yes it is just one extra item to cycle through, but that one extra key press takes time, 

There's a fundamental flaw in the reasoning there. Initial focus should be somewhere decent. That's it. That's the desired result. Nothing more, nothing less.

Also note that #12328 looks nothing like this. While it smells like something that should be a user patch it's gated behind `Device:hasSymKey() or Device:hasScreenKB()`. Hopefully other Kindle users don't share my opinion on it, because the simple fact is *other Kindle users added it in the first place*.

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:278]: Yes, that was added to kindle specifically, albeit before anyone considered the more logical approach of mapping things to buttons. And i am sure nobody is going to be upset because they now need to press fewer buttons.

Now as for hasDPad and not useDPad, tht is very much what i am seeking counsel on, it doesn’t really need to be on kindle, as it is already there through the key event. So what is the best way of reaching everyone yet not adding it to where it should not go.

Comment (Frenzie) [reply frontend/ui/widget/textviewer.lua:278]: > And i am sure nobody is going to be upset because they now need to press fewer buttons and in turn gain improvements in performance.

I am. I think it tanks usability having to press a menu button instead of just happily navigating. But no matter, don't you simply want `hasDPad and not (Device:hasSymKey() or Device:hasScreenKB())`?

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:278]: > I am. I think it tanks usability having to press a menu button instead of just happily navigating.

yes, I would agree with this if only those old kindles had newer and faster processors but alas they don't and that "free navigation" comes at a cost. I wish you could see it in person and then you would perhaps understand.

> But no matter, don't you simply want hasDPad and not (Device:hasSymKey() or Device:hasScreenKB())

that is perhaps what I want, in trying to be clever (or helpful) I might be overtaking this whole matter.

Comment (Commodore64user) [inline frontend/ui/widget/textviewer.lua:278]:
```suggestion
    if Device:hasDPad() and not (Device:hasSymKey() or Device:hasScreenKB()) then
```

File: frontend/ui/widget/textviewer.lua [hunk 2]
```diff
@@ -272,6 +274,16 @@ function TextViewer:init(reinit)
...
 
+    -- NT: add titlebar.left_button (hamburger menu) to FocusManager.
+    if Device:hasDPad() and not (Device:hasSymKey() or Device:hasScreenKB()) then
+        -- ButtonTable calls refocusWidget on init, but we'll mangle the layout,
+        -- so kill the initial highlight while FocusManager can still find the current focused item...
+        self.button_table:handleEvent(Event:new("Unfocus"))
+        table.insert(self.button_table.layout, 1, { self.titlebar.left_button })
+        -- And refocus manually on the *actual* layout
+        self.button_table:refocusWidget()
+    end
```
Comment (poire-z) [inline frontend/ui/widget/textviewer.lua:285]: > another stray widget that needs its menu mapped to Menu key. borrowing code from @NiLuJe to fix focusmanager shenanigans.

What was the previous behaviour ?
Aren't we doing this kind of things elsewhere (or is it just focusmanager that does that?) - and if yes, why didn't we need this unfocus/refocus?
And if it is just focusmanager that does that, wouldn't it be better if it was still only it, ie. by adding/using some method?
Not really familiar with all that, so just pinging @NiLuJe to confirm this is ok.

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:285]: Remember how we once discussed always showing focus, not just after pressing a button? Well we are fixing that now, by we I mean @NiLuJe, in any case we must make sure focus is always accurate now. BTW i stole this from him ;)

Comment (Commodore64user) [reply frontend/ui/widget/textviewer.lua:285]: As to why we need to unfocus refocus see https://github.com/koreader/koreader/pull/12361#issuecomment-2293177294 dictionary one is exactly the same case as this.

Comment (NiLuJe) [reply frontend/ui/widget/textviewer.lua:285]: > Aren't we doing this kind of things elsewhere (or is it just focusmanager that does that?) - and if yes, why didn't we need this unfocus/refocus?
> And if it is just focusmanager that does that, wouldn't it be better if it was still only it, ie. by adding/using some method?
> Not really familiar with all that, so just pinging @NiLuJe to confirm this is ok.

Like the comment says, FocusManager has a method to do it (refocusWidget), and ButtonDialog calls int (on... init? for some reason, probably because a lot of stuff besides instantiation calls init I'd wager).

But the issue is that for the unfocus that happens in there to work, FocusManager has to be able to *find* the widget at the right coords in self.layout.

But here, we're completely mangling self.layout outside of FocusManager, so it completely breaks that.

Hence: manual unfocus => mangle layout => refocus

Comment (poire-z) [reply frontend/ui/widget/textviewer.lua:285]: > Hence: manual unfocus => mangle layout => refocus

Ok, fine with me if fine with you.

Comment (NiLuJe) [reply frontend/ui/widget/textviewer.lua:285]: Yup, see the DQL commit in #12361 (which is where that comment originated)
